boa/src/builtins/value/operations.rs

```diff
@@ -1,4 +1,72 @@
 use super::*;
+use crate::Interpreter;
+use std::borrow::Borrow;
+
+#[allow(clippy::float_cmp)]
+/// https://tc39.es/ecma262/#sec-numeric-types-number-equal
+fn strict_number_equals<T: Into<f64>>(a: T, b: T) -> bool {
+    let a: f64 = a.into();
+    let b: f64 = b.into();
+
+    if a.is_nan() || b.is_nan() {
+        return false;
+    }
+
+    a == b
+}
+
+impl Value {
+    // https://tc39.es/ecma262/#sec-strict-equality-comparison
+    pub fn strict_equals(&self, other: &Self) -> bool {
+        if self.get_type() != other.get_type() {
+            return false;
+        }
+
+        if self.is_number() {
+            return strict_number_equals(self, other)
+        }
+
+        same_value_non_number(self, other)
+    }
+
+    // https://tc39.es/ecma262/#sec-abstract-equality-comparison
+    pub fn equals(&mut self, other: &mut Self, interpreter: &mut Interpreter) -> bool {
+        if self.get_type() == other.get_type() {
+            return self.strict_equals(other);
+        }
+
+        match (self.data(), other.data()) {
+            _ if self.is_null_or_undefined() && other.is_null_or_undefined() => true,
+
+            // https://github.com/rust-lang/rust/issues/54883
+            (ValueData::Integer(_), ValueData::String(_))
+            | (ValueData::Rational(_), ValueData::String(_))
+            | (ValueData::String(_), ValueData::Integer(_))
+            | (ValueData::String(_), ValueData::Rational(_))
+            | (ValueData::Rational(_), ValueData::Boolean(_))
+            | (ValueData::Integer(_), ValueData::Boolean(_)) => {
+                let a: &Value = self.borrow();
+                let b: &Value = other.borrow();
+                strict_number_equals(a, b)
+            }
+            (ValueData::Boolean(_), _) => {
+                other.equals(&mut Value::from(self.to_integer()), interpreter)
+            }
+            (_, ValueData::Boolean(_)) => {
+                self.equals(&mut Value::from(other.to_integer()), interpreter)
+            }
+            (ValueData::Object(_), _) => {
+                let mut primitive = interpreter.to_primitive(self, None);
+                primitive.equals(other, interpreter)
+            }
+            (_, ValueData::Object(_)) => {
+                let mut primitive = interpreter.to_primitive(other, None);
+                primitive.equals(self, interpreter)
+            }
+            _ => false,
+        }
+    }
+}
 
 impl PartialEq for Value {
     fn eq(&self, other: &Self) -> bool {
```

Comment:
> Maybe the `PartialEq` trait should be strict equality, instead of abstract equality.
> 
> What do you think @Razican?

Yep, `PartialEq` is a Rust trait, so it should do a "Rust equality" which is a strict equality.

Reply:
I don't change it cause many places to use it. I will look at this.

Reply:
remove `PartialEq`?

Reply:
That's another option, yes, to just remove this implementation, since `Value` is a JS value, and doesn't have much sense to use a Rust operation to compare a JS value.

Reply:
done